<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px"> 

## Lab: SQL Practice

_Author: Matt Brems_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [1]:
# Import pandas and create_engine from `sqlalchemy`

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data
engine = \
create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [2]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT *
FROM stores
"""

store = pd.read_sql_query(sql, engine)

**Prompt 1 Answer**:

In [3]:
store.shape[0]

1973

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [4]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT *
FROM stores
ORDER BY name
"""

pd.read_sql_query(sql, engine)

store                                name store_status  \
0      4986  218 Fuel Express & Chubby's Liquor            A   
1      3047          3047 Cub Foods / Iowa City            I   
2      3060               3060 Cub Foods / Ames            I   
3      3061         3061 Cub Foods / Sioux City            I   
4      4845                  3rd St Convenience            A   
...     ...                                 ...          ...   
1968   3992       Xo Food And Liquor / Waterloo            I   
1969   4066                       Xpress Liquor            I   
1970   4886                      Yr Dollar Stop            A   
1971   4908                Zapf's Pronto Market            A   
1972   4874                      Z's Quickbreak            A   

                                          store_address address_info  
0     68 Monroe St\nFloyd, IA 50435\n(43.12249790500...         None  
1     855 Highway 1 West\nIowa City, IA 522460000\n(...         None  
2     3121 Grand Ave\nAmes, IA 500100000\n(42.053458...         None  
3     1732 Hamilton Blvd\nSioux City, IA 511030000\n...         None  
4     216, 3rd St Se\nCedar Rapids, IA 52401\n(41.97...         None  
...                                                 ...          ...  
1968  428 Franklin St\nWaterloo, IA 507030000\n(42.5...         None  
1969  1201 North Jefferson Ste 100\nIndianola, IA 50...         None  
1970  1500 1st Ave Ne\nCedar Rapids, IA 52402\n(41.9...         None  
1971  107 1st St Sw\nElkader, IA 52043\n(42.85350191...         None  
1972  201, E Washington St\nMount Pleasant, IA 52641...         None  

[1973 rows x 5 columns]

**Prompt 2 Answer**: 3061 Cub Foods / Sioux City is the fourth alphabetical order store.

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [5]:
## Check sales columns and values
sql = """
SELECT *
FROM sales
LIMIT 1
"""

pd.read_sql_query(sql, engine)

date convenience_store  store county_number county category  \
0 2014-11-13              None   2635            82  Scott  1032200   

           category_name vendor_no                            vendor   item  \
0  IMPORTED VODKA - MISC       370  Pernod Ricard USA/Austin Nichols  34036   

     description  pack  liter_size state_btl_cost btl_price  bottle_qty  total  
0  Absolut Pears    12         750         $11.49    $17.24           4  68.96

In [6]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT *
FROM products
WHERE category_name = 'IMPORTED VODKA'
ORDER BY vendor_name
LIMIT 3
"""

pd.read_sql_query(sql, engine)

item_no   category_name        item_description  vendor  \
0   903883  IMPORTED VODKA  Luksusowa Potato Vodka      14   
1   904783  IMPORTED VODKA            Vesica Vodka      14   
2   903387  IMPORTED VODKA           Iceberg Vodka      16   

                 vendor_name  bottle_size  pack  inner_pack   age proof  \
0  Adamba Imports Int'l Inc.         1750     6           1  None    80   
1  Adamba Imports Int'l Inc.          750    12           1  None    80   
2            Admiral Imports         1000    12           1  None    80   

   list_date   upc             scc bottle_price  shelf_price  case_cost  
0 2003-07-01  None            None       $14.12        21.18      84.70  
1 2010-06-14  None  10076800000000        $5.99         8.99      71.88  
2 2009-01-20  None            None       $10.00        15.00     120.00

In [21]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT p.category_name, store, s.vendor, s.btl_price, bottle_qty
FROM sales as s
JOIN products as p
on s.item = p.item_no
WHERE p.category_name = 'IMPORTED VODKA'
ORDER BY vendor_name
LIMIT 3
"""

top3 = pd.read_sql_query(sql, engine)
top3

category_name  store                     vendor btl_price  bottle_qty
0  IMPORTED VODKA   4509  Adamba Imports Int'l Inc.     $8.99          12
1  IMPORTED VODKA   2548      A Hardy / U.S.A. Ltd.    $28.50           6
2  IMPORTED VODKA   3773      A Hardy / U.S.A. Ltd.    $28.50           3

**Prompt 3 Answer**: 
1.  Store 4509, Vendor: Adamba Imports Int'l Inc. Sold 12 bottles at $8.99 each

2.  Store 2548, Vendor: A Hardy / U.S.A. Ltd..    Sold 6 bottles at $28.50 each

3.  Store 3773, Vendor: A Hardy / U.S.A. Ltd..    Sold 3 bottles at $28.50 each

### Prompt 4

What is the total amount of liquor sold in February 2015?

In [20]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT SUM(total)::MONEY
FROM sales as s
WHERE DATE_PART('month',date) = '02' AND DATE_PART('year',date) = '2015'
"""
pd.read_sql_query(sql, engine)

sum
0  $21,295,350.55

**Prompt 4 Answer**: $21,295,350.55 -> ~21 millions for all stores sales combined


### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [25]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT vendor, COUNT(*) AS total_transactions, SUM(bottle_qty) AS total_bottle_sold, SUM(total)::MONEY AS total_revenue
FROM sales as s
WHERE category_name = 'IMPORTED VODKA'
GROUP BY vendor
ORDER BY 2 DESC
"""

pd.read_sql_query(sql, engine)

vendor  total_transactions  \
0       Pernod Ricard USA/Austin Nichols               42338   
1                    Bacardi U.S.A. Inc.               26441   
2                        Diageo Americas               15725   
3        Constellation Wine Company Inc.               12565   
4                        Jim Beam Brands                5850   
5                            Stoli Group                4443   
6                   Bacardi U.S.A., Inc.                3344   
7                   Imperial Brands Inc.                2799   
8                 Moet Hennessy USA Inc.                2153   
9       Constellation Wine Company, Inc.                1672   
10                               Proximo                1628   
11                               MHW Ltd                1159   
12                        Luxco-St Louis                1128   
13                   Wilson Daniels Ltd.                 815   
14                 Imperial Brands, Inc.                 426   
15            Russian Standard Vodka USA                 415   
16                      Quadro Group LLC                 289   
17              Brown-Forman Corporation                 263   
18               Moet Hennessy USA, Inc.                 256   
19           WILLIAM GRANT AND SONS INC.                 232   
20                      Klin Spirits LLC                 218   
21                                 LUXCO                 140   
22           Russian Standard Vodka, USA                  71   
23                                  None                  62   
24          WILLIAM GRANT AND SONS, INC.                  56   
25                 A Hardy / U.S.A. Ltd.                  33   
26           W. J. Deutsch And Sons Ltd.                  24   
27                     Jinro America Inc                  21   
28              Firestarter Spirits Inc.                  20   
29      Western Spirits Beverage Co. LLC                  18   
30                             TY KU LLC                  17   
31          Park Street Imports LLC Voli                  12   
32  Shaw Ross International Importers LL                  10   
33                      Sazerac Co. Inc.                  10   
34                     Klin Spirits, LLC                   5   
35            The Patron Spirits Company                   4   
36                     Wine Intelligence                   3   
37      Stanley Stawski Distributing Co.                   2   
38        Charles Jacquin Et. Cie., Inc.                   2   
39                     Phenix Brands LLC                   2   
40                 Ti Beverage Group Ltd                   2   
41                A Hardy / U.S.A., Ltd.                   1   
42             Adamba Imports Int'l Inc.                   1   

    total_bottle_sold  total_revenue  
0              403450  $7,880,365.88  
1              236047  $5,771,242.07  
2              152038  $3,346,425.42  
3              116053  $1,895,005.51  
4              104119  $1,807,097.23  
5               35513    $832,071.19  
6               30239    $765,347.31  
7               21033    $265,317.79  
8               10966    $323,126.54  
9               12956    $200,283.64  
10               9431    $178,212.03  
11               4493    $109,190.48  
12               8933    $116,914.19  
13               2396     $84,246.72  
14               3054     $37,950.20  
15               2421     $40,361.58  
16               1054     $25,928.40  
17               1730     $23,337.88  
18               1032     $30,783.53  
19               1119     $20,981.25  
20                784     $15,483.84  
21               2374     $32,425.32  
22                360      $5,400.00  
23                368      $6,009.44  
24                252      $4,725.00  
25                320      $9,120.00  
26                768      $8,593.92  
27               1056      $7,319.04  
28                 54        $994.14  
29                484      $9,453.60  
30      

**Prompt 5 Answer**: Pernod Ricard USA/Austin Nichols has the most transactions.

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [9]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT (btl_price - state_btl_cost) AS Markup
FROM sales
"""

markup = pd.read_sql_query(sql, engine)
markup

markup
0         $5.75
1         $5.75
2         $5.75
3         $8.58
4         $2.50
...         ...
3049908   $5.75
3049909  $10.00
3049910   $5.75
3049911   $5.75
3049912  $10.00

[3049913 rows x 1 columns]

In [ ]:
# sql = """
# SELECT *
# FROM sales
# """

# df = pd.read_sql_query(sql, engine)
# df

In [ ]:
## Alternatively Select all column and add new column in dataframe
# df["Markup"] = df["btl_price"] - df["state_btl_cost"]

In [ ]:
# df

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [20]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT store, AVG(markup/state_btl_cost)*100::DECIMAL AS percentage_markup
FROM (SELECT store, btl_price, state_btl_cost, (btl_price - state_btl_cost) AS markup
FROM sales
GROUP BY store, btl_price, state_btl_cost) AS temp
GROUP BY store
ORDER BY 2 DESC
"""

pd.read_sql_query(sql, engine)

store  percentage_markup
0      3859          57.843149
1      3934          57.771609
2      3833          57.602838
3      9001          57.597214
4      4013          57.169854
...     ...                ...
1347   4858          50.005467
1348   9018          50.000000
1349   4784          49.999887
1350   4534          49.980544
1351   4693          49.924824

[1352 rows x 2 columns]

**Prompt 7 Answer**: Store 3859 have the highest average percentage markup at 57.84%

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [7]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT temp.store, AVG(markup/state_btl_cost)*100::DECIMAL AS avg_percentage_markup,
MIN(markup/state_btl_cost)*100 AS min_percentage_markup,
MAX(markup/state_btl_cost)*100 AS max_percentage_markup,
name, store_address
FROM (SELECT store, btl_price, state_btl_cost, (btl_price - state_btl_cost) AS markup
FROM sales
GROUP BY store, btl_price, state_btl_cost) AS temp
JOIN stores
ON temp.store = stores.store
GROUP BY temp.store, name, store_address
ORDER BY 2 DESC
LIMIT 5
"""

pd.read_sql_query(sql, engine)

store  avg_percentage_markup  min_percentage_markup  max_percentage_markup  \
0   3859              57.843149              49.846154             670.876289   
1   3934              57.771609              49.846154             670.876289   
2   3833              57.602838              49.846154             670.876289   
3   9001              57.597214              49.943757             209.278351   
4   4013              57.169854              49.857143             670.876289   

                           name  \
0  Wal-Mart 0750 / Independence   
1     Wal-Mart 1509 / Maquoketa   
2      Wal-Mart 3394 / Atlantic   
3         Cedar Ridge Vineyards   
4        Wal-Mart 0841 / Tipton   

                                       store_address  
0  302 Enterprise Dr Sw\nIndependence, IA 5064400...  
1  103 E Carlisle\nMaquoketa, IA 520600000\n(42.0...  
2  1905 East 7th St\nAtlantic, IA 500220000\n(41....  
3  1441 Marak Rd Nw\nSwisher, IA 523380000\n(41.8...  
4  1126 Highway 38 North\nTipton, IA 527720000\n(...

**Prompt 8 Answer**: 
1. Wal-Mart 0750 / Independence
2. Wal-Mart 1509 / Maquoketa
3. Wal-Mart 3394 / Atlantic
4. Cedar Ridge Vineyards
5. Wal-Mart 0841 / Tipton

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery).

In [23]:
sql = """
SELECT *
FROM sales as s
LIMIT 1
"""

pd.read_sql_query(sql, engine)

date convenience_store  store county_number county category  \
0 2014-11-13              None   2635            82  Scott  1032200   

           category_name vendor_no                            vendor   item  \
0  IMPORTED VODKA - MISC       370  Pernod Ricard USA/Austin Nichols  34036   

     description  pack  liter_size state_btl_cost btl_price  bottle_qty  total  
0  Absolut Pears    12         750         $11.49    $17.24           4  68.96

In [25]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.county, (SUM(bottle_qty)::DECIMAL/population::DECIMAL) AS bottle_per_capita, 
AVG(liter_size) AS average_bottle_size
FROM sales as s
JOIN counties as c
ON s.county = c.county
GROUP BY s.county,population
ORDER BY 2 DESC
LIMIT 9
"""

pd.read_sql_query(sql, engine)

county  bottle_per_capita  average_bottle_size
0      Dickinson          22.262615           974.184437
1     Black Hawk          15.390159           860.966660
2           Polk          15.086295           891.874028
3    Cerro Gordo          14.126656           986.401569
4          Scott          13.910340           869.916897
5        Johnson          13.166883           895.198259
6           Linn          12.833984           885.064695
7  Pottawattamie          11.581292           890.907515
8        Kossuth          11.379656          1040.529457

**Prompt 9 Answer**: Kossuth have the highest average bottle size.

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [4]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT *
FROM (SELECT btl_price, bottle_qty, total::MONEY, (btl_price*bottle_qty) AS total2
FROM sales) AS temp
WHERE total != total2
"""

pd.read_sql_query(sql, engine)

Empty DataFrame
Columns: [btl_price, bottle_qty, total, total2]
Index: []

In [5]:
sql = """
SELECT btl_price, bottle_qty, total::MONEY, (btl_price*bottle_qty) AS total2
FROM sales
"""

total = pd.read_sql_query(sql, engine)

In [6]:
total[total["total"] != total["total2"]]

Empty DataFrame
Columns: [btl_price, bottle_qty, total, total2]
Index: []

In [7]:
total.head()

btl_price  bottle_qty    total   total2
0    $17.24           4   $68.96   $68.96
1    $17.24          12  $206.88  $206.88
2    $17.24          12  $206.88  $206.88
3    $25.73           6  $154.38  $154.38
4     $7.50           6   $45.00   $45.00

**Prompt 10 Answer**: 0 (No) Transaction have a value of total that is not equal to btl_price time bottle_qty.

## Additional Questions

●    How many total products are in the Products table?

In [17]:
sql = """
SELECT *
FROM products
"""

products = pd.read_sql_query(sql, engine)
products.head(1)

item_no                       category_name        item_description  \
0   904616  MISC. AMERICAN CORDIALS & LIQUEURS  Travis Hasse Apple Pie   

   vendor vendor_name  bottle_size  pack  inner_pack   age proof  list_date  \
0     305     Mhw Ltd          750    12           1  None    40 2009-02-11   

    upc   scc bottle_price  shelf_price  case_cost  
0  None  None        $9.77        14.66     117.22

In [12]:
products.shape[0]

9977

●    Who are the top most diverse vendors (i.e. they have the highest number of distinct products)? 
     How many different products do they have?


In [18]:
sql = """
SELECT vendor, vendor_name, COUNT (DISTINCT item_no) AS distinct_items, COUNT (category_name) AS distinct_items
FROM products
GROUP BY vendor, vendor_name
ORDER BY 3 DESC
"""

vendor = pd.read_sql_query(sql, engine)
vendor

vendor                       vendor_name  count  count
0        65                   Jim Beam Brands    925    916
1       260                   Diageo Americas    907    898
2       370  Pernod Ricard Usa/austin Nichols    599    597
3       885         Yahara Bay Distillers Inc    579    578
4       259     Heaven Hill Distilleries Inc.    388    385
..      ...                               ...    ...    ...
266     104                    Asdspirits Llc      1      1
267     129             Devotion Spirits Inc.      1      1
268      30         Austin Nichols & Co. Inc.      1      1
269      89      Temperance Distiling Company      1      1
270     138     Bom Dia Imports Dba Novo Fogo      1      1

[271 rows x 4 columns]

In [ ]:
●    Which products sell the best by total number of unit sales? (Hint: you will need to find a way of doing a cross-join between the sales and product tables.)

●    Which products sell the best by total dollar value of sales?

●    What are the top 10 categories of liquor sold based on the total amount of sales revenue?

●    Which rum products have sales greater than $10,000? How about whiskey or vodka products?

●    Which county sold the most amount of vodka during February 2014?

●    Which counties were in the top 10 counties for vodka sales in any month in 2014?

●    Create a report that shows how many times a county appeared in the “top 10 counties for vodka sales in a month” list over the course of 2014.

●    What is the trend of sales by month? Break up variables such as bottle_price or liter_size into categories (for example: cheap, medium, or expensive).

●    Which stores sell one of the top five most expensive bottles of alcohol?

●    How many stores have more than $2,000,000 in total sales?

●    How many stores have an average bottle price greater than $20?

●    Which stores have the highest sales of items over 90 proof